django/db/models/query_utils.py

```diff
@@ -110,6 +114,31 @@ def flatten(self):
             else:
                 yield child
 
+    def check(self, against, using=DEFAULT_DB_ALIAS):
+        """
+        Do a database query to check if the expressions of the Q instance
+        matches against the expressions.
+        """
+        # Avoid circular imports.
+        from django.db.models import Value
+        from django.db.models.sql import Query
+        from django.db.models.sql.constants import SINGLE
+
+        query = Query(None)
+        for name, value in against.items():
+            if not hasattr(value, "resolve_expression"):
+                value = Value(value)
+            query.add_annotation(value, name, select=False)
+        query.add_annotation(Value(1), "_check")
+        # This will raise a FieldError if a field is missing in "against".
+        query.add_q(self)
+        compiler = query.get_compiler(using=using)
+        try:
+            return compiler.execute_sql(SINGLE) is not None
+        except DatabaseError as e:
+            logger.warning("Got a database error calling check() on %r: %s", self, e)
```

Comment:
@felixxm Do you happen to know why we do log instead of failing loudly? In https://code.djangoproject.com/ticket/34840#comment:11 the failures are actually database errors because it cannot properly check for NULL without casting but due to the `return True` below the error is hidden.

Reply:
iirc it was to deal with problematic backends like Oracle

Reply:
The tests seem to suggest that it is due to `RawSQL` https://github.com/django/django/pull/15659/files#diff-4e19e10988d9113c1c0ada7c38d6f3bba1a9a8629da429e34816682342998e0dR244 -- but if that is the case I'd still think an error would make sense instead of returning wrong data.

Reply:
It has nothing to do with Oracle, check out the [discussion](https://github.com/django/django/pull/14625#pullrequestreview-921674376).

Reply:
Right, sadly this also masked the errors while developing support for psycopg3. Could we just skip the checks for `RawSQL` at all instead of trying/hoping for a database error?

Reply:
@felixxm Just FYI in case this is pertinent to this discussion: I wasn't aware of the earlier discussions on RawSQL but I was sure we talked about logging being beneficial to skip Oracle's quirks when you Simon & I were talking about handling NULL values. I can't find the conversation but this is the PR: https://github.com/django/django/pull/16038 🤷‍♂️